File: src/core/jwt.ts
/** Decodes unverified JWT display/routing claims; none of these values may gate authorization. */
import { decodeBase64url } from "./pkce.js";

const CLAIM_NAMESPACE = "https://api.openai.com/auth";

export interface UnverifiedClaims {
  accountId?: string;
  planType?: string;
  email?: string;
}

function decodePayload(token: string): Record<string, unknown> | null {
  const payload = token.split(".")[1];
  if (payload === undefined) return null;
  try {
    const parsed: unknown = JSON.parse(new TextDecoder().decode(decodeBase64url(payload)));
    return parsed !== null && typeof parsed === "object" ? parsed as Record<string, unknown> : null;
  } catch {
    return null;
  }
}

/** The returned claims are unverified and suitable only for display and request routing. */
export function extractUnverifiedClaims(idToken?: string, accessToken?: string): UnverifiedClaims {
  const payload = decodePayload(idToken ?? "") ?? decodePayload(accessToken ?? "") ?? {};
  const namespace = payload[CLAIM_NAMESPACE];
  const auth = namespace !== null && typeof namespace === "object" ? namespace as Record<string, unknown> : {};
  const claims: UnverifiedClaims = {};
  if (typeof auth.chatgpt_account_id === "string") claims.accountId = auth.chatgpt_account_id;
  if (typeof auth.chatgpt_plan_type === "string") claims.planType = auth.chatgpt_plan_type;
  if (typeof payload.email === "string") claims.email = payload.email;
  return claims;
}

File: python/src/chatgpt_oauth/jwt.py
"""Unverified JWT metadata extraction for display and request routing only."""

import json
from dataclasses import dataclass
from typing import cast

from .pkce import decode_base64url

_CLAIM_NAMESPACE = "https://api.openai.com/auth"


@dataclass(frozen=True, slots=True)
class UnverifiedClaims:
    """Carries unverified display/routing claims that must never authorize access."""

    account_id: str | None = None
    plan_type: str | None = None
    email: str | None = None


def _payload(token: str) -> object:
    try:
        segment = token.split(".")[1]
        return cast(object, json.loads(decode_base64url(segment).decode()))
    except (IndexError, UnicodeDecodeError, ValueError):
        return None


def extract_unverified_claims(
    id_token: str | None = None, access_token: str | None = None
) -> UnverifiedClaims:
    """Extracts display/routing claims only; these unverified claims never authorize."""
    payload = _payload(id_token or "")
    if not isinstance(payload, dict):
        payload = _payload(access_token or "")
    if not isinstance(payload, dict):
        return UnverifiedClaims()
    namespace = payload.get(_CLAIM_NAMESPACE)
    auth = namespace if isinstance(namespace, dict) else {}
    account_id = auth.get("chatgpt_account_id")
    plan_type = auth.get("chatgpt_plan_type")
    email = payload.get("email")
    return UnverifiedClaims(
        account_id=account_id if isinstance(account_id, str) else None,
        plan_type=plan_type if isinstance(plan_type, str) else None,
        email=email if isinstance(email, str) else None,
    )
